Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -0,0 +1,354 @@
+    tl.debug("Searching for files to upload");
+
+    try {
+        const rootFolderStats = tl.stats(ftpOptions.rootFolder);
+        if (rootFolderStats.isFile()) {
+            const file = ftpOptions.rootFolder;
+            tl.debug(file + " is a file. Ignoring all file patterns");
+            return [file];
+        }
+
+        const allFiles = tl.find(ftpOptions.rootFolder);
+
+        // filePatterns is a multiline input containing glob patterns
+        tl.debug(
+            "searching for files using: " +
+            ftpOptions.filePatterns.length +
+            " filePatterns: " +
+            ftpOptions.filePatterns
+        );
+
+        // minimatch options
+        const matchOptions = { matchBase: true, dot: true } as MatchOptions;
+        const platform = tl.getPlatform()
+        tl.debug("Platform: " + platform);
+        if (platform === tl.Platform.Windows) {
+            matchOptions["nocase"] = true;
+        }
+
+        tl.debug("Candidates found for match: " + allFiles.length);
+        for (let i = 0; i < allFiles.length; i++) {
+            tl.debug("file: " + allFiles[i]);
+        }
+
+        // use a set to avoid duplicates
+        let matchingFilesSet: Set<string> = new Set();
+
+        for (let i = 0; i < ftpOptions.filePatterns.length; i++) {
+            let normalizedPattern: string = path.join(
+                ftpOptions.rootFolder,
+                path.normalize(ftpOptions.filePatterns[i])
+            );
+
+            tl.debug("searching for files, pattern: " + normalizedPattern);
+
+            const matched = tl.match(allFiles, normalizedPattern, undefined, matchOptions);
+            tl.debug("Found total matches: " + matched.length);
+            // ensure each result is only added once
+            for (let j = 0; j < matched.length; j++) {
+                let match = path.normalize(matched[j]);
+                let stats = tl.stats(match);
+                if (!ftpOptions.preservePaths && stats.isDirectory()) {
+                    // if not preserving paths, skip all directories
+                } else if (matchingFilesSet.add(match)) {
+                    tl.debug(
+                        "adding " + (stats.isFile() ? "file:   " : "folder: ") + match
+                    );
+                    if (stats.isFile() && ftpOptions.preservePaths) {
+                        // if preservePaths, make sure the parent directory is also included
+                        let parent = path.normalize(path.dirname(match));
+                        if (matchingFilesSet.add(parent)) {
+                            tl.debug("adding folder: " + parent);
+                        }
+                    }
+                }
+            }
+        }
+
+        return Array.from(matchingFilesSet).sort();
+    }
+    catch (err) {
+        tl.error(err);
+        tl.setResult(tl.TaskResult.Failed, tl.loc("UploadFailed"));
+    }
+
+    return [];
+}
+
+async function run() {
```

Comment:
While we're rewriting the task, I think it would be a good time to break the components up into separate files. This will make it easier to find code and make the connections easier to see through exports.

I think this would be a good breakdown:
1. `ftpuploadtask.ts`: `findFiles()`, `run()`
2. `retry.ts`: `sleep()`, *`retryWithNewClient()`
3. `ftp.ts`: `getAccessOption()`, *`getFtpClient()`, *`FtpOptions`,
4. `progresstracker.ts`: *`ProgressTracker`

\* = export

Reply:
This is another thing that I agree in principle for a more involved project, but I don't think the tradeoff is worth the trouble in simple task scripts.  I don't like the proliferation of files, especially after working with Java for many years...  :)   In the current file, I think it may make sense to break off `ProgressTracker` into its own module, but then editing this task with vanilla vim requires open multiple files. Unless we are reusing the module, I don't really think it make the code more maintainable given the meat of this script is only ~300 lines.